test/es-module/test-esm-detect-ambiguous.mjs

```diff
@@ -185,7 +175,7 @@ describe('--experimental-detect-module', { concurrency: !process.env.TEST_PARALL
       ]);
 
       strictEqual(stderr, '');
-      strictEqual(stdout, 'null\nexecuted\n');
+      strictEqual(stdout, 'undefined\nexecuted\n');
```

Comment:
What about this change?

Reply:
Yes, this is where the code paths diverging was an issue. On current `main`, without passing `detect-module`, this prints `commonjs`. So as part of unflagging, this needs to change to either `null` or `undefined`, and I think `undefined` makes more sense as a new value.

This code path was returning `null`: https://github.com/nodejs/node/blob/7d07c95f551f200424bdefd68aed15c6ee6d7570/lib/internal/modules/esm/get_format.js#L116-L118

And this was returning `undefined`:
https://github.com/nodejs/node/blob/7d07c95f551f200424bdefd68aed15c6ee6d7570/lib/internal/modules/esm/get_format.js#L173-L174

I think it’s a mistake to return differing values based on which code path gets followed. This is what https://github.com/nodejs/node/pull/53015 intended to fix, but you didn’t want to make a breaking change without a reason. Now that this is changing anyway, I think this is the time?

It wasn’t my intent in this PR to fix this; I wasn’t thinking about `null` at all when I put this together. But if not now, I’m not sure when we could ever fix this.

Reply:
tbh I don't think it's a problem if we never fix it. That being said, I think it'd be better to discuss it separately, if we're not happy with `null` being returned, we should have it landed separately, so it can be backported/reverted/bisected on its own.

Reply:
How about we fix it as two commits in this same PR, that we land via commit-queue-rebase? That way they get released together and we minimize disruption.

I think we _do_ need to fix this, because the next PR I want to do is to remove `default-type` and you can see it’s all over this section of the code. At some point the refactoring here is going to make it difficult if not impossible to preserve the cases that are supposed to be `null` versus the ones that are supposed to be `undefined`, so I’d rather just simplify it so that we don’t have further issues in refactoring later.

Likewise, I think the tests should check just for `undefined` so we don’t accidentally introduce `null` in the future and cause this problem all over again. I haven’t checked, but I wouldn’t be surprised if a too-permissive test (or a lack of one) is how we got here, where one of us wrote code that returned `undefined` and the other of us wrote `null` without realizing.

Reply:
OK for having the refactor in the same PR, but the breaking change should still be a separate PR.

> Likewise, I think the tests should check just for `undefined` so we don’t accidentally introduce `null` in the future and cause this problem all over again.

I disagree it's going to help us, if we want to validate the returned value, we should right a test for that and use `assert.strictEqual`. If we're writing a loader, let's write a loader, one that our users can take inspiration from, not one that's half testing if the returned value is what we expect.